src/meta-srv/src/region/lease_keeper.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod mito;
+pub mod utils;
+
+use std::collections::{HashMap, HashSet};
+
+use common_meta::key::TableMetadataManagerRef;
+use snafu::ResultExt;
+use store_api::storage::{RegionId, TableId};
+
+use self::mito::retain_active_regions;
+use crate::error::{self, Result};
+
+/// Region Lease Keeper removes any inactive [RegionRole::Leader] regions.
+pub struct RegionLeaseKeeper {
+    table_metadata_manager: TableMetadataManagerRef,
+}
+
+impl RegionLeaseKeeper {
+    pub fn new(table_metadata_manager: TableMetadataManagerRef) -> Self {
+        Self {
+            table_metadata_manager,
+        }
+    }
+}
+
+impl RegionLeaseKeeper {
+    /// Retains active [RegionRole::Leader] regions, returns inactive regions.
+    ///
+    /// - It grants new lease if the `datanode_id` is the corresponding leader peer in `region_routes`.
+    /// - It removes a leader region if the `datanode_id` isn't the corresponding leader peer in `region_routes`.
+    ///     - Expected as [RegionRole::Follower] regions.
+    ///     - Unexpected [RegionRole::Leader] regions.
+    /// - It removes a region if the region's table metadata is not found.
+    pub async fn retain_active_regions(
```

Comment:
It burns my brain when staring these codes trying to sort out a clean logic. Can you refactor the codes to make them simpler? I think the codes are made complex because the functions have multiple desires. For example, `retain_active_regions` is finding inactive regions and filtering the datanode regions at the same time.